objects/src/notes/note_execution_hint.rs

```diff
@@ -0,0 +1,182 @@
+    /// 1152..1279, 2176..2303 etc.
+    OnBlockSlot {
+        epoch_len: u8,
+        slot_len: u8,
+        slot_offset: u8,
+    },
+}
+
+impl NoteExecutionHint {
+    /// Creates a [NoteExecutionHint::None] variant
+    pub fn none() -> Self {
+        NoteExecutionHint::None
+    }
+
+    /// Creates a [NoteExecutionHint::Always] variant
+    pub fn always() -> Self {
+        NoteExecutionHint::Always
+    }
+
+    /// Creates a [NoteExecutionHint::AfterBlock] variant based on the given `block_num`
+    pub fn after_block(block_num: u32) -> Self {
+        NoteExecutionHint::AfterBlock { block_num }
+    }
+
+    /// Creates a [NoteExecutionHint::OnBlockSlot] for the given parameters
+    pub fn on_block_slot(epoch_len: u8, slot_len: u8, slot_offset: u8) -> Self {
+        NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset }
+    }
+
+    /// Returns whether the note execution conditions validate for the given `block_num`
+    pub fn can_be_consumed(&self, block_num: u32) -> bool {
+        match self {
+            NoteExecutionHint::None => false,
+            NoteExecutionHint::Always => true,
+            NoteExecutionHint::AfterBlock { block_num: hint_block_num } => {
+                block_num >= *hint_block_num
+            },
+            NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset } => {
+                let epoch_len_blocks: u32 = 1 << epoch_len;
+                let slot_len_blocks: u32 = 1 << slot_len;
+
+                let block_epoch_index = block_num / epoch_len_blocks;
+
+                let slot_start_block =
+                    block_epoch_index * epoch_len_blocks + (*slot_offset as u32) * slot_len_blocks;
+                let slot_end_block = slot_start_block + slot_len_blocks;
+
+                block_num >= slot_start_block && block_num < slot_end_block
+            },
+        }
+    }
+}
+
+// CONVERSIONS
+// ================================================================================================
+
+impl From<NoteExecutionHint> for u64 {
+    fn from(hint: NoteExecutionHint) -> Self {
+        match hint {
+            NoteExecutionHint::None => 0,
+            NoteExecutionHint::Always => 1 << 32,
+            NoteExecutionHint::AfterBlock { block_num } => {
+                let tag = 2;
+                (tag << 32) | (block_num as u64)
+            },
+            NoteExecutionHint::OnBlockSlot { epoch_len, slot_len, slot_offset } => {
+                let tag: u64 = 3;
+                let payload: u64 = (((epoch_len as u32) << 16)
+                    | ((slot_len as u32) << 8)
+                    | (slot_offset as u32)) as u64;
+                (tag << 32) | payload
+            },
+        }
+    }
+}
+
+impl TryFrom<u64> for NoteExecutionHint {
+    type Error = NoteError;
+    fn try_from(encoded: u64) -> Result<Self, NoteError> {
+        let tag = ((encoded >> 32) & 0xFF) as u8;
```

Comment:
This ignores the upper bits of the `u64` value. Are we intentionally discarding these? If so, I'd probably document that we don't care about the upper bits. But also, I wonder if it would be safer to make sure that upper bits are zeros.